pallets/asset/src/lib.rs

```diff
@@ -2176,6 +2197,35 @@ impl<T: Trait> Module<T> {
         (transfer_status, weight_for_valid_transfer)
     }
 
+    /// Sets the initial total supply of a confidential asset.
+    /// Only called by the token owner of a confidential asset. Can be called only once.
+    ///
+    /// # Arguments
+    /// * `origin` Secondary key of the token owner.
+    /// * `ticker` Ticker of the token.
+    /// * `total_supply` Ticker of the token.
+    pub fn unchecked_set_total_supply(
+        did: IdentityId,
+        ticker: Ticker,
+        total_supply: T::Balance,
+    ) -> DispatchResult {
+        ensure!(Self::is_owner(&ticker, did), Error::<T>::Unauthorized);
+        // Read the token details
+        let mut token = Self::token_details(&ticker);
+        ensure!(
+            token.total_supply == Zero::zero(),
+            Error::<T>::CanSetTotalSupplyOnlyOnce
+        );
+        ensure!(
+            total_supply != Zero::zero(),
+            Error::<T>::TotalSupplyMustBePositive
+        );
+        token.total_supply = total_supply;
+        <Tokens<T>>::insert(&ticker, token);
+        Self::deposit_event(RawEvent::TotalSupplyChanged(did, ticker, total_supply));
```

Comment:
and I think we don't even need the `TotalSupplyChanged` event here, We can emit an `Issued` event here for the owner did.